### ISSUE TITLE
[jacobsen on 2007-11-20 18:11:28] : eventBuilder - fileDispatcher code busy waits 100% of available CPU

### ISSUE THREAD

**Issue by dglo** [dglo/pdaq#638]
No associated GitHub commit

**Comment by dglo** [comment]
[dglo on 2007-11-20 18:23:45]
Looking at FileDispatcher.java, I notice that it does 'if (tempFile.length() > maxFileSize)' after every payload write, which is probably the source of the stat64 calls.  You could easily keep a running byte count, adding buffer.limit() after each event, instead of checking length() each time.  You'd probably want to initialize the byte count with tempFile.length() when you first open the 

John didn't note that he was doing a 1-string run with EB running on hub29 and all other components on hub21, and the hub29 CPU was still pegged at 100% usage ... 10% user and 90% system.

I think we're doing around 600 Hz ... I can see where doing 600 stat calls per second might increase the system load a bit :-)

I've got no idea, however, what's causing all the gettimeofday() calls

**Comment by dglo** [comment]
[jacobsen on 2007-11-20 18:48:33]
Stat probably calls gettimeofday(), perhaps to compare file times with the current wallclock time for some reason...

**Comment by dglo** [comment]
[ksbeattie on 2007-11-20 22:31:45]
A quick and dirty test commenting out the 'if (tempFile.length() > maxFileSize)' block showed that this didn't drop the cpu from being pegged, which means that this isn't the cause.  The next conjecture is that the EB itself needs some profiling to see where the culprit is.  I've created issue 2037 to make Dave's suggested change to save the tempFile.length() calls.

**Comment by dglo** [comment]
[jacobsen on 2007-11-26 16:24:19]
From: 	  dave.glowacki@icecube.wisc.edu
	Subject: 	[Daq-dev] Event builder odds'n'ends
	Date: 	November 21, 2007 4:38:06 PM CST
	To: 	  daq-dev@icecube.wisc.edu

[Keith and John, we could probably track down the ichub29 "polling" thread using the technique outlined below...]

I spent a little time today trying to profile the event builder (because I didn't want to start another big trigger optimization push right before a 4-day break, and because it seemed like a "fun" thing to do :-)

In order to profile it, I needed to run the Sun JVM.  Doing so, I ran into our old friend, the locked-up JVM.  I used 'top' to find the looping thread (which can be seen by using the 'H' command to enable the per-thread display) and then used strace to watch that thread repeatedly call sched_yield().  Some googling turned up the fact that the Linux Java thread display contains the thread PID (it's the 'nid'), and that revealed that the looping thread was the "VM Thread".

I started a new event builder, pointed strace at the VM Thread, then started up a run, and saw the thread flip over fram what looked like normal behavior to the sched_yield() lockup.

I filed a bug report, but since I didn't really have any way for them to reproduce the bug, I don't have much hope that they'll do anything about it.

The bug report form asked me if I had tried running with -Xint (which disables HotSpot), and that worked around the problem, so I was able to do a couple of profiling runs.

As you might expect, the bulk of the time was taken up almost equally by input and output.  I don't think there's much waste there, it's just that reading and writing data takes a lot of time.

Most of the remaining time is eaten up by building the event payload, and only a tiny percent is spent in FileDispatcher (which Keith will be happy to hear :-)

I didn't see the "polling" thread which John noticed on ichub29, but maybe we'd be able to track it down by matching the thread ID to the thread-dump output...

These results again show that we could probably improve performance of the system by overhauling the payload system.  It's a non-trivial job, however, so I'm certainly not volunteering for it before April 2008...
